File: Longest_Palindrome_Substring.cpp
class Solution {
public:
    /*
    Idea: 2D Dynamic Programming
    */
    string longestPalindrome(string s) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        int len = s.length();
        if (len <= 1) return s;
        bool pal[len][len];
        for (int n = 0; n < len; n ++) {
            pal[n][n] = true;
        }
        
        int maxLen = 1, maxStart = 0;
        for (int i = 0; i < len; i ++){
            for (int j = i; j >= 0; j --){
                if (s[j] == s[i] && (i - j < 2 || pal[j+1][i-1])) {
                    pal[j][i] = true;
                    if (i - j + 1 > maxLen) {
                        maxLen = (i - j + 1);
                        maxStart = j;
                    }
                }
                else {pal[j][i] = false;}
            }
        }
        return s.substr(maxStart, maxLen);
    }
};

File: python/Longest_Palindrome_Substring.py
class Solution:
    # @return a string
    def longestPalindrome(self, s):
        n = len(s)
        if (n == 0): return s
        maxLen = 0
        maxStart = 0
        ptr1 = 0
        ptr2 = 0
        for i in range(n):
            ptr1 = i - 1
            ptr2 = i + 1
            while ptr1 >= 0 and ptr2 < n and s[ptr1] == s[ptr2]:
                ptr1 = ptr1 - 1
                ptr2 = ptr2 + 1
            if ptr2-ptr1-1 >= maxLen:
                maxLen = ptr2-ptr1-1
                maxStart = ptr1 + 1
            ptr1 = i
            ptr2 = i+1
            while ptr1 >= 0 and ptr2 < n and s[ptr1] == s[ptr2]:
                ptr1 = ptr1 - 1
                ptr2 = ptr2 + 1
            if ptr2-ptr1-1 >= maxLen:
                maxLen = ptr2-ptr1-1
                maxStart = ptr1 + 1
        return s[maxStart: maxStart+maxLen]
